### ISSUE TITLE
Path parameter cannot be optional is not accurate anymore

### ISSUE THREAD

**Issue by timotheeguerin** [microsoft/typespec#4126]
With the introduction of uri template path parameter should be allowed to be optional at least in certain conditions

[Playground Link](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7Cgp1c2luZyBUeXBlU3BlYy5IdHRwOwoKQHJvdXRlKCJvcHRpb25hbHsvcGFyYW19L2xpc3QiKQpvcCDIGyjFGj86IHN0cmluZyk6IHZvaWQ7Cg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%22linterRuleSet%22%3A%7B%22extends%22%3A%5B%22%40typespec%2Fhttp%2Fall%22%5D%7D%7D)

```tsp
// Allowed to be optional
@route("optional{/param}/list")
op optional(@path param?: string): void;
// param: foo => optional/foo/list
// param: undefined => optional/list

// Not allowed 
@route("optional/{param}/list")
op optional(@path param?: string): void;
                               ^ error should include the / in the interpolation to omit the segment

// Allow in autoRoute
@autoRoute
op optional(@path param?: string): void;
```

To watch for in the autoroute case it should generate the url with the `/` inside of the `{}`

**Comment by jhendrixMSFT** [comment]
KeyVault has some operations that have an optional path parameter (a key's version). At present, there's no way to describe this in tsp.

**Comment by gearama** [comment]
E.G. 

/**
* The GET operation is applicable to any secret stored in Azure Key Vault. This operation requires the secrets/get permission.
*/
#suppress "@azure-tools/typespec-azure-core/use-standard-operations" "Foundations.Operation is necessary for Key Vault"
@summary("Get a specified secret from a given key vault.")
@route("/secrets/{secret-name}/{secret-version}")
@get
op getSecret is KeyVaultOperation<
  {
    /**
     * The name of the secret.
     */
    @path("secret-name")
    @clientName("name", "go")
    secretName: string;
 
    /**
     * The version of the secret. This URI fragment is optional. If not specified, the latest version of the secret is returned.
     */
    @path("secret-version")
    @clientName("version", "go")
    secretVersion: string;
  },
  SecretBundle
>;

Both of these
kv/secretName/secretVersion?api-version=...
and 
kv/secretName?api-version=
should be valid

**Comment by antkmsft** [comment]
Technically, I believe secret-version could be an empty string, but that alone will not help us to generate code with API where secretVersion is not a required argument to the method which calls this REST API.

**Comment by allenjzhang** [comment]
Blocking C++ & Rust for KeyVault

**Comment by timotheeguerin** [comment]
updated the issue with what we should do, please reply if thats not enough

**Comment by jhendrixMSFT** [comment]
So, in order for this to work, one has to apply the `?` to the param and specify the path segment's start as `{/` yes? It seems reasonable to me.

For [KV secrets](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/keyvault/Security.KeyVault.Secrets/routes.tsp#L65-L94) we'd end up with something like this then?

```
@route("/secrets/{secret-name}{/secret-version}")
@patch
op updateSecret is KeyVaultOperation<
  {
    @path("secret-name")
    @clientName("name", "go")
    secretName: string;

    @path("secret-version")
    @clientName("version", "go")
    secretVersion?: string;

    #suppress "deprecated" "Property flattening is supported for legacy scenarios like Key Vault's"
    @body
    @flattenProperty
    parameters: SecretUpdateParameters;
  },
  SecretBundle
>;
```

**Comment by timotheeguerin** [comment]
indeed, the `?` is to mark the parameter optional for the typespec type, `{/abc}` is how you make it to drop the `/` in uri template specs
